packages/yourloops/pages/profile/profil-page-context.hook.ts

```diff
@@ -0,0 +1,159 @@
+  const isUserHcp = user.isUserHcp()
+
+  const [profileForm, setProfileForm] = useState<ProfileForm>({
+    birthday: user.birthday,
+    birthPlace: user.profile?.patient?.birthPlace ?? '',
+    feedbackAccepted: !!user?.profile?.contactConsent?.isAccepted,
+    firstName: user.firstName,
+    hcpProfession: user.profile?.hcpProfession ?? HcpProfession.empty,
+    ins: user.profile?.patient?.ins ?? undefined,
+    lang: user.preferences?.displayLanguageCode ?? getCurrentLang(),
+    lastName: user.lastName,
+    referringDoctor: user.profile?.patient?.referringDoctor ?? undefined,
+    sex: user.profile?.patient?.sex ?? undefined,
+    ssn: user.profile?.patient?.ssn ?? undefined,
+    units: user.settings?.units?.bg ?? Units.gram
+  })
+  const [saving, setSaving] = useState<boolean>(false)
+
+  const errors: ProfileErrors = {
+    birthday: isUserPatient && !REGEX_BIRTHDATE.test(profileForm.birthday),
+    firstName: !profileForm.firstName,
+    hcpProfession: isUserHcp && profileForm.hcpProfession === HcpProfession.empty,
+    ins: isUserPatient && (!profileForm.ins || (profileForm.ins && profileForm.ins.length !== 15)),
+    lastName: !profileForm.lastName,
+    ssn: isUserPatient && (!profileForm.ssn || (profileForm.ssn && profileForm.ssn.length !== 15))
+  }
+
+  const updatedProfile = useMemo<Profile>(() => {
+    const profile: Profile = {
+      ...user.profile,
+      firstName: profileForm.firstName,
+      lastName: profileForm.lastName,
+      fullName: `${profileForm.firstName} ${profileForm.lastName}`
+    }
+
+    if (isUserPatient) {
+      profile.patient = {
+        ...profile.patient,
+        birthday: profileForm.birthday,
+        birthPlace: profileForm.birthPlace,
+        ins: profileForm.ins,
+        sex: profileForm.sex,
+        ssn: profileForm.ssn,
+        referringDoctor: profileForm.referringDoctor
+      }
+    }
+
+    if (user.isUserHcp()) {
+      profile.hcpProfession = profileForm.hcpProfession
+    }
+
+    if (isUserHcp && !!user?.profile?.contactConsent?.isAccepted !== profileForm.feedbackAccepted) {
+      profile.contactConsent = {
+        isAccepted: profileForm.feedbackAccepted,
+        acceptanceTimestamp: new Date().toISOString()
+      }
+    }
+
+    return profile
+  }, [isUserHcp, isUserPatient, profileForm.birthPlace, profileForm.birthday, profileForm.feedbackAccepted, profileForm.firstName, profileForm.hcpProfession, profileForm.ins, profileForm.lastName, profileForm.referringDoctor, profileForm.sex, profileForm.ssn, user])
+
+  const updatedSettings: Settings = {
+    ...user.settings,
+    units: { bg: profileForm.units }
+  }
+  const updatedPreferences: Preferences = {
+    ...user.preferences,
+    displayLanguageCode: profileForm.lang
+  }
+
+  const profileChanged: boolean = !isEqual(user.profile, updatedProfile)
+  const preferencesChanged: boolean = !isEqual(user.preferences, updatedPreferences)
+  const settingsChanged: boolean = !isEqual(user.settings, updatedSettings)
```

Comment:
profileChanged preferencesChanged settingsChanged are only used in the method saveProfile. I feel like they should be initialized in it.

Reply:
No, they're also used to compute canSave variable, which is exposed by the hook for components.